cosmwasm/packages/wasmi-runtime/src/wasm/runtime/contract.rs

```diff
@@ -469,8 +452,47 @@ impl WasmiApi for ContractInstance {
     }
 
     // stub, for now
-    fn query_chain_index(&mut self, _query_ptr_ptr: i32) -> Result<Option<RuntimeValue>, Trap> {
-        Err(WasmEngineError::NotImplemented.into())
+    fn query_chain_index(&mut self, query_ptr_ptr: i32) -> Result<Option<RuntimeValue>, Trap> {
+        let query_buffer = self.extract_vector(query_ptr_ptr as u32).map_err(|err| {
+            error!("query_chain() error while trying to read canonical address from wasm memory",);
+            err
+        })?;
+
+        trace!(
+            "query_chain() was called from WASM code with {:?}",
+            String::from_utf8_lossy(&query_buffer)
+        );
+
+        // Call query_chain (this bubbles up to x/compute via ocalls and FFI to Go code)
+        // Returns the value from x/compute
+        let (answer, gas_used) = encrypt_and_query_chain(
+            &query_buffer,
+            &self.context,
+            self.user_nonce,
+            self.user_public_key,
+        )?;
+        trace!(
+            "query_chain() got answer from outside with result {:?}",
+            String::from_utf8_lossy(&answer)
+        );
+        trace!(
+            "query_chain() got answer from outside with gas {}",
+            gas_used
+        );
+
+        self.use_gas_externally(gas_used)?;
+
+        let ptr_to_region_in_wasm_vm =   self.write_to_memory(&answer)
```

Comment:
extra spaces here.

_I swear I can solve captchas._

Reply:
It's the effing rustfmt and macros

Reply:
I figured :joy: